### ISSUE TITLE
Introduce the concept of state transitions in Blazor to allow for e.g. animations

### ISSUE THREAD

**Issue by tomRedox** [aspnet/AspNetCore#10389]
### Is your feature request related to a problem? Please describe.
Following on from [my StackOverflow question](https://stackoverflow.com/questions/56219984/how-to-animate-state-transitions-in-blazor)...

In my Blazor components I often render components based on either a conditional statement, e.g.

```
@if (_contact.IsCustomer)
{
    <SalesOrdersList Customer="@_contact" />
}
```

Or from a loop, e.g.

```
@foreach(var salesOrder in _customer.SalesOrders)
{
    <SalesOrderSummary SalesOrder="@salesOrder" />
}
```

When I change the state I'd like to animate the state transition so that the components fade in/out. In the examples above that might happen when IsCustomer changes, or when a record is added or removed from the SalesOrders collection.

Animating adding a component can be done with CSS, but I think that animating the removal of a component will be impossible because it simply doesn't exist in the next render, so there's nothing to hang the animation onto?

### Describe the solution you'd like
In React there used to be a series of attributes related to state transitions, which now seem to have been replaced by [react-transition-group](https://github.com/reactjs/react-transition-group).  That looks like a big hammer for what seems like a fairly small nut.

Ideally I'd favour something really simple to use, like having an attribute for a CSS class that's implemented and allowed to complete prior to the item being removed from the render tree, but I appreciate that may be completely impractical.

There are potentially lots of other things you might want to happen in a state transition too, some that can't be handled with CSS alone too, but I think the animation side of things is fairly important to aid users in understanding the consequences of a particular UI interaction.

**Comment by mkArtakMSFT** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by rxelizondo** [comment]
Perhaps this is implied in the original post, but another thing that would be really nice to have animated would be page transitions.

Essentially, it would be nice that as you navigate from page to page, the new pages would swoop in and out from the left, right, up or down depending on some setting. I am very new to Blazor so I am not sure if this is possible to do right now but I have a feeling it is not.

**Comment by dazinator** [comment]
This is something that other client side frameworks such as @SteveSandersonMS's knockoutjs made very easy, and as much as am liking blazor, and am aware it's still premature, I am surprised Blazor has made it this far without any such mechanism yet provided. Is anyone able to share an example of doing a simple fadeout / fadein transition of one fragement that is being removed from the dom, and a new fragement that is being added to the dom?

**Comment by dazinator** [comment]
I settled on the following solution to applying a transition to an element before removal, it's not the greatest because I am using `Task.Delay` with a specified time in milliseconds which needs to match up o the duration of the transition specified in the css:

`Transition.razor` component;
```
<div class="@(ToggleActive ? ToggleTransitionOnCssClassName: ToggleTransitionOffCssClassName)">
    @ChildContent;
</div>

@code {

[Parameter] RenderFragment ChildContent { get; set; }

[Parameter] string ToggleTransitionOnCssClassName { get; set; } = "";
[Parameter] string ToggleTransitionOffCssClassName { get; set; } = "";

[Parameter] int TransitionDurationMilliseconds { get; set; } = 200;

public bool ToggleActive { get; set; }

[Parameter] EventCallback<bool> TransitionEnded { get; set; }

public async Task ToggleTransition()
{
    ToggleActive = !ToggleActive;
    await Task.Delay(TransitionDurationMilliseconds);
    await TransitionEnded.InvokeAsync(ToggleActive);
}


}

```

and it's used like so from a parent page or component:

```

       @if (!RenderThingy)
        {
            <Transition @ref="Transition" TransitionDurationMilliseconds="500" ToggleTransitionOnCssClassName="m-fadeOut" ToggleTransitionOffCssClassName="m-fadeIn" TransitionEnded="@TransitionComplete">
                <RenderThingy OnDismissed="@OnDismissed"></RenderThingy>
            </Transition>
        }

@code {

    Transition Transition { get; set; }

    async Task OnDismissed()
    {
        await Transition.ToggleTransition();
    }
    private void TransitionComplete(bool toggleState)
    {
        RenderThingy = false;
    }
}

```
and css:

```
.m-fadeIn {
    visibility: visible;
    opacity: 1;
    animation: fadein 500ms;
}

@keyframes fadein {
    from {
        opacity: 0;
    }

    to {
        opacity: 1;
    }
}

.m-fadeOut {
    animation: fadeout 500ms;
}

@keyframes fadeout {
    from {
        opacity: 1;
    }

    to {
        opacity: 0;       
    }
}

```

**Comment by Weezeal** [comment]
@tomRedox Did you manage to get anywhere with this? I'm currently building a chart where I want items to animate between state changes. I'm currently getting around it by toggling a boolean as a parameter on each component I want to animate (quite similarly to @dazinator 's suggestion above). However, this is pretty awful, as it results in basically everything re-rendering.

**Comment by dazinator** [comment]
I hate my "solution".. blazor needs something proper for this!

**Comment by tomRedox** [comment]
@Weezeal no I'm afraid not, from experience with React I don't think there really is anywhere to go until the framework incorporates something we can use for this.

The [stack overflow question](https://stackoverflow.com/questions/56219984/how-to-animate-state-transitions-in-blazor) is at 600 views and 1 upvote in 4 months, so I guess there's not that much demand for this yet, but I agree with @dazinator that this is one of those things that clients and managers have got so used to seeing in apps built with other SPA frameworks that it's now slightly jarring when an app doesn't do it; it's the sort of thing that non-technical people spot.

**Comment by SteveSandersonMS** [comment]
The solution I'd expect is to create an animator component that renders either a list or a single item, and incorporates logic to delay the removal of each item so it can be animated out. Blazor already has good primitives for templates components, so the resulting APIs should be pretty nice. This would be essentially the same solution that is used in other SPA frameworks.

This is achievable in user code and doesn't require a built-in framework feature. I'm not saying it's easy, but hopefully someone in the community will find time to do it. It's something I may do myself at some point but have other priorities in the short term.

**Comment by dazinator** [comment]
@SteveSandersonMS 
> , and incorporates logic to delay the removal of each item so it can be animated out. 

Not sure if this was the best approach but I tried this but I don't like it because suppose you are animating a fade out transition, in your css you might specify a value such as 2 seconds to fade over. In blazor you must now task.delay the removal by atleast 2 seconds otherwise the UI element will be removed before the animation completes. The animation values are held in css, yet they are now tightly coupled with blazor parameters where the delays specified must allt tie up. Of you decide to change the css to fade out over 3 seconds  you must find every usage of your blazor animation component that renders this fade out and change its delay value to 3.
It's working but there has to be a better solution, one that uses events in the DOM to work out when an animation has complete perhaps so it can remove after that event is raised and  ot before. That way you wouldn't need to specify any delay value.

**Comment by SteveSandersonMS** [comment]
Yes, you can implement any strategy you want for deciding how long the delay should be. It could delay until some DOM event occurs.

**Comment by dazinator** [comment]
Ok thanks. I'll try and revisit this at some point then and see if I can get past that final hurdle. If it works out I'll open source something and post a link here.

**Comment by dazinator** [comment]
Here is my solution: https://github.com/dazinator/BlazorDeferredRemove

**Comment by VaclavElias** [comment]
Another solution https://github.com/mikoskinen/Blazor.Animate

**Comment by NVentimiglia** [comment]
This is my number one blocker for adopting blazor. I need to know that I can keep my application beautiful with intro, outro, update, and list-move transition animations. I really enjoy the solution that Vue.JS came up with. It solves all my perceived use cases in a very non-obtrusive way. The solution is much like Blazor.Animate in that we wrap the html in a <Animate> component. 

https://vuejs.org/v2/guide/transitions.html

https://github.com/mikoskinen/Blazor.Animate/issues/10

**Comment by Dedigraz** [comment]
I don't know whether this issue is still actively being researched, but I believe there's some great progress in the [shared element transition proposal](https://github.com/WICG/shared-element-transitions) that the blazor team could look at.

**Comment by 0xakihiko** [comment]
I'm really surprised Blazor doesn't have this already honestly, it seems like a staple in most libraries.

**Comment by MarkusRodler** [comment]
Maybe the upcoming Shared Element Transition API could help: https://developer.chrome.com/blog/shared-element-transitions-for-spas/

**Comment by oliverw** [comment]
"We've moved this issue to the Backlog milestone." - This is turning into "Groundhog Day" 😄

**Comment by mkArtakMSFT** [comment]
The View Transitions API seems promising. Unfortunately, it's not yet supported by all major browsers: https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API#browser_compatibility

**Comment by MarkusRodler** [comment]
> The View Transitions API seems promising. Unfortunately, it's not yet supported by all major browsers: https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API#browser_compatibility

Yeah I know, Internet Explorer 11 support is not that good 😉
Joking aside, would it still be an option to implement the feature with it, even if the rest of the browsers still have to follow suit? In principle, it would be better than nothing.

**Comment by SteveSandersonMS** [comment]
For people who are interested in using View Transitions, have a look at this: https://github.com/jsakamoto/Toolbelt.Blazor.ViewTransition

**Comment by dimitri-dev** [comment]
Sadly, view transitions API don't seem to be supported for Safari on iOS. I consider this topic to be a really valuable one from the perspective of Blazor + MAUI Hybrid apps. Will there be any progress made in regards to this?

**Comment by MarkusRodler** [comment]
> Sadly, view transitions API don't seem to be supported for Safari on iOS. I consider this topic to be a really valuable one from the perspective of Blazor + MAUI Hybrid apps. Will there be any progress made in regards to this?

It should be supported since 18.0? https://caniuse.com/mdn-api_viewtransition

**Comment by dimitri-dev** [comment]
@MarkusRodler I agree that it's looking a bit better now than it was back in November.

So, latest metrics show 68% iOS 18 adoption for all Apple devices, 76% if we count only the last 4 years worth of models.
That is according to the App Store Metrics from Jan 21 2025 => https://developer.apple.com/support/app-store/

But adoption in the first few months was not nearly as high (many bugs with iOS 18 and posts urging people not to update).
Building an app that would feel/look good for 20-25% of the iOS users makes little-to-no sense?

Even now, roughly 1 out of 4 people using your app with a device released **in the past 4 years** (that's iPhone 13 and above) may not have the experience you intended. That's a no-deal for me.
